### ISSUE TITLE
PSP Files For Missing Elements

### ISSUE THREAD

**Issue by umbriquse** [JuliaMolSim/DFTK.jl#397]
Just found that some elements are not available for use in the list_pspfile() function. Would it be possible to use a psp file found online? I came across a series of [psp files that were generated using the Troullier-Martins method](https://www.abinit.org/sites/default/files/PrevAtomicData/psp-links/lda_tm.html). I'm curious how appropriate they would be in regards to simulation, and if they could be incorporated into the DFTK database of elements.

**Comment by antoine-levitt** [comment]
We only support Goedecker pseudopotentials right now; supporting others is https://github.com/JuliaMolSim/DFTK.jl/issues/2. It's not especially hard, one just needs to write a parser and add a simple radial integration routine to accomodate numerical rather than analytical radial parts, but we never got around to it. Regarding accuracy, I think the old Troullier-Martins pseudopotentials are pretty bad, and the more modern ONCVPSP pseudos are better, but I don't have much experience with pseudopotentials so couldn't say for sure.

**Comment by umbriquse** [comment]
Ok, so what I plan on doing is to create a function(s) that parses the psp files that were generated by the Troullier-Martins method. Should I base this function on the "parse_hgh_file()" function? Maybe even use type dispatch to differentiate the two?

**Comment by antoine-levitt** [comment]
That'd be awesome! So there are three parts to this : first, create a PspTM (or perhaps PspNormConserving since there's nothing specific to TM) which implements the same methods as PspHgh (we might then create an AbstractPsp type from which these two would derive but I'd rather wait until we can better see what the interface should be). Once this is done you should be able to create it manually and test it. The second part is writing a wrapper that reads a psp file and creates a PspTM struct. This is a bit more tricky because there are many formats around. Here I would search the internet for the most convenient format (pspnc, XML, upsp...) and implement that. I believe there are converters around. The third part is figuring out an "external" API and changing the list/load psp functions ; that we should do at the end.

Let us know if you need any help!

**Comment by mfherbst** [comment]
Closing this in favour of the more specific #2 and #134. Feel free to add specific issues for the other mentioned file formats to keep track.

Regarding availability of pseudo definitions, I think psp8 and UPF are promising as ABINIT and QE have tons of files for a large list of elements / functionals.

**Comment by antoine-levitt** [comment]
So from the formula side I think what you need is to compute the integral of a radial function specified on a grid against a bessel function j(q r), for several q. How do people do that, just simple sums? No special tricks eg for high q?

**Comment by mfherbst** [comment]
No idea I never checked, but I can tell you that for small Ecuts you get noticable differences between our analytical GTHs and the radial grid ones used e.g. by ABINIT, so probably there is not too much cleverness in this.
